rfc-176-replacing-static.md

```diff
@@ -0,0 +1,127 @@
+
+#### How will we do this in practice?
+Although sprockets-rails wants to put things in one directory (the assets.prefix directory configured in application.rb), it is possible to split assets by source by modifying the Sprockets::Asset class to return a digest path with different prefixes based on the asset name. Since asset engine gems tend to be namespaced (govuk_publishing_components is, for instance), we can use the gem name to determine whether a thing should go into shared or not. If the asset starts with a gem name on a list we provide, it can be prefixed with /shared, and if not it can be prefixed with the application name as usual.
+
+Ideally we’d do this by subclassing the Sprockets environment (which is responsible for loading Assets) to provide a modified class, but at worst we could do it by monkeypatching (there’s a proof of concept of this).
+
+![Proposed asset deploy pathway](rfc-176/proposed_govuk_asset_pathway.png)
+
+### Layouts (move to `govuk_web_layout`)
+All of the layouts in `static` are banner-aware configuration wrappers around a call to the public layout component included in the [govuk_publishing_components] gem. In practise this means that applications are having to make network calls to ask another application to return them a layout that they already have access to, and could easily configure themselves if necessary.
+
+We propose that to start with we replicate this layout configuration in the gem (so that there is little additional configuration needed in the host application), and at a later stage we audit the layouts to determine if the additional configuration is something that can be provided in a default load-time step and explicitly overridden by the applications. This should give us more flexibility in terms of personalisation.
+
+### Header and Footer Content (move to `govuk_web_layout`)
+Updating header and footer content is a developer task in `static` - there are no content items that can be edited in the publishing apps. Ideally we’d prefer this not to be the case, but that’s out of the scope of this current document.
+
+We propose for the moment moving the header and footer content into the gem, where it can be included as part of the layouts also provided by the gem. This gives us a similar developer experience to today (in that editing the content is in one place), while allowing us to remove that functionality from [static].
+
+### Additional Considerations
+
+Adding a couple more gems to every frontend app may mean that frontend development may become more awkward in situations where local versions of multiple gems are needed. These are likely to be few, but we should be prepared to monitor the developer experience in these situations. Hopefully, though, the clear division of purpose between these gems and the components gem should mean that most work can be done with at most one local gem.
```

Comment:
This is an area of concern, so I appreciate it being highlighted here. I'm happy with this approach for now on the basis that if it doesn't work we're able to make changes in future.

I think it's worth pointing out that `govuk_publishing_components` has had a lot of effort put into it that I think we should replicate in some of these new gems, particularly if they're going to have existing components transferred across. For example component conventions and principles, automated accessibility testing, file name conventions testing (and probably some others I can't recall right now). I guess there's also the question of the component guide - will the layout gem have its own guide to show the layout components, will that be weird having two virtually identical component guides at different URLs, etc.

Reply:
I don't see the components necessarily moved across at this stage. There's maybe an argument for it (in that the intervention and emergency banner components are currently _only_ used in banners, so could be moved to the banners gem with little upset), but those steps don't get us any special benefits in the same way that moving the banner handling/test code out of static does, so I haven't included it as part of this proposal.

Reply:
That seems sensible, thanks.